#### Importing Required Libraries: 

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import os
import pymatgen.core as pg
from sklearn.feature_selection import VarianceThreshold
import matplotlib.pyplot as plt 
import seaborn as sns
from sklearn.model_selection import train_test_split

#### Importing Training Data (various Features of Alloys)

In [2]:
df = pd.read_excel('Data/Training/Features.xlsx') ## Import the training dataset
df

Compound  Temperature(K)  mean_EffectiveCoordination  \
0           AlAs       152.91209                    5.720212   
1            AlP       140.72993                    5.368858   
2    AlP (F-43m)       162.26119                    5.720212   
3           AlSb       148.62687                    5.720212   
4          B2AsP       142.34014                    5.731569   
..           ...             ...                         ...   
114         TePb       147.79412                    6.000000   
115       Ti2SnC       100.70368                    9.383733   
116         TlBr       160.52725                   11.956922   
117         TlCl       160.26508                   11.956922   
118          TlI       148.12925                   11.956922   

     var_EffectiveCoordination  min_EffectiveCoordination  \
0                 3.108624e-15                   5.720212   
1                 3.150000e-10                   5.368858   
2                 1.330000e-15                   5.720212   
3                 0.000000e+00                   5.720212   
4                 1.021271e-01                   5.541343   
..                         ...                        ...   
114               4.440000e-16                   6.000000   
115               1.691867e+00                   6.000000   
116               8.880000e-16                  11.956922   
117               8.880000e-16                  11.956922   
118               8.880000e-16                  11.956922   

     max_EffectiveCoordination  var_MeanBondLength  min_MeanBondLength  \
0                     5.720212        2.775558e-16            1.000000   
1                     5.368858        1.660000e-11            1.000000   
2                     5.720212        5.000000e-16            1.000000   
3                     5.720212        2.780000e-16            1.000000   
4                     5.935824        1.160746e-02            0.977966   
..                         ...                 ...                 ...   
114                   6.000000        0.000000e+00            1.000000   
115                  11.858700        9.280932e-02            0.814381   
116                  11.956922        1.670000e-16            1.000000   
117                  11.956922        2.220000e-16            1.000000   
118                  11.956922        2.220000e-16            1.000000   

     max_MeanBondLength  mean_BondLengthVariation  ...  max_SpaceGroupNumber  \
0              1.000000                  0.160958  ...                  29.5   
1              1.000000                  0.146248  ...                 225.0   
2              1.000000                  0.160958  ...                 225.0   
3              1.000000                  0.160958  ...                 166.0   
4              1.023215                  0.160397  ...                 229.0   
..                  ...                       ...  ...                   ...   
114            1.000000                  0.000000  ...                 166.0   
115            1.163322                  0.089257  ...                 194.0   
116            1.000000                  0.051982  ...                 166.0   
117            1.000000                  0.051982  ...                 166.0   
118            1.000000                  0.051982  ...                  64.0   

     min_SpaceGroupNumber  most_SpaceGroupNumber  frac_sValence  \
0                     225                  166.0     195.500000   
1                       2                  113.5       0.500000   
2                     166                  195.5       0.222222   
3                       2                  166.0       0.307692   
4                     166                  229.0       0.233333   
..                    ...                    ...            ...   
114                     2                  166.0       0.307692   
115                   141                  194.0       0.307692   
116                     2                   84.0     

In [3]:
print('Shape of DataFrame is : ',df.shape)
print('Null entries in DataFrame are : ', df.isnull().sum().sum())
df.dropna(inplace=True)
print('After dropping null values, shape of DataFrame becomes : ',df.shape)

Shape of DataFrame is :  (119, 273)
Null entries in DataFrame are :  1
After dropping null values, shape of DataFrame becomes :  (118, 273)


#### Importing LTC vs Temperature Data:

In [4]:
Y = pd.read_excel('Data/Training/Initial Dataset.xlsx')
Y

Compound        LTC  Temperature(K)
0        AlAs  266.67806       152.91209
1        AlAs  220.88175       159.37363
2        AlAs  175.08544       185.21978
3        AlAs  129.28913       211.06593
4        AlAs  106.39098       243.37363
...       ...        ...             ...
2140      TlI    0.05528       751.19048
2141      TlI    0.04856       818.19728
2142      TlI    0.04184       885.20408
2143      TlI    0.04184       952.21088
2144      TlI    0.04184      1019.21769

[2145 rows x 3 columns]

#### Making combined training dataset: Merging LTC data at various temperature with features of alloys

In [5]:
df.drop(['Temperature(K)'],axis = 1, inplace=True)
data_full = pd.merge(df,Y,on='Compound')
data_full

Compound  mean_EffectiveCoordination  var_EffectiveCoordination  \
0         AlP                    5.368858               3.150000e-10   
1         AlP                    5.368858               3.150000e-10   
2         AlP                    5.368858               3.150000e-10   
3         AlP                    5.368858               3.150000e-10   
4         AlP                    5.368858               3.150000e-10   
...       ...                         ...                        ...   
2123      TlI                   11.956922               8.880000e-16   
2124      TlI                   11.956922               8.880000e-16   
2125      TlI                   11.956922               8.880000e-16   
2126      TlI                   11.956922               8.880000e-16   
2127      TlI                   11.956922               8.880000e-16   

      min_EffectiveCoordination  max_EffectiveCoordination  \
0                      5.368858                   5.368858   
1                      5.368858                   5.368858   
2                      5.368858                   5.368858   
3                      5.368858                   5.368858   
4                      5.368858                   5.368858   
...                         ...                        ...   
2123                  11.956922                  11.956922   
2124                  11.956922                  11.956922   
2125                  11.956922                  11.956922   
2126                  11.956922                  11.956922   
2127                  11.956922                  11.956922   

      var_MeanBondLength  min_MeanBondLength  max_MeanBondLength  \
0           1.660000e-11                 1.0                 1.0   
1           1.660000e-11                 1.0                 1.0   
2           1.660000e-11                 1.0                 1.0   
3           1.660000e-11                 1.0                 1.0   
4           1.660000e-11                 1.0                 1.0   
...                  ...                 ...                 ...   
2123        2.220000e-16                 1.0                 1.0   
2124        2.220000e-16                 1.0                 1.0   
2125        2.220000e-16                 1.0                 1.0   
2126        2.220000e-16                 1.0                 1.0   
2127        2.220000e-16                 1.0                 1.0   

      mean_BondLengthVariation  var_BondLengthVariation  ...  \
0                     0.146248             2.090000e-11  ...   
1                     0.146248             2.090000e-11  ...   
2                     0.146248             2.090000e-11  ...   
3                     0.146248             2.090000e-11  ...   
4                     0.146248             2.090000e-11  ...   
...                        ...                      ...  ...   
2123                  0.051982             9.710000e-17  ...   
2124                  0.051982             9.710000e-17  ...   
2125                  0.051982             9.710000e-17  ...   
2126                  0.051982             9.710000e-17  ...   
2127                  0.051982             9.710000e-17  ...   

      most_SpaceGroupNumber  frac_sValence  frac_pValence  frac_dValence  \
0                     113.5       0.500000       0.500000       0.000000   
1                     113.5       0.500000       0.500000       0.000000   
2                     113.5       0.500000       0.500000       0.000000   
3                     113.5       0.500000       0.500000       0.000000   
4                     113.5       0.500000       0.500000       0.000000   
...                     ...            ...            ...            ...   
2123                   33.0       0.222222       0.222222       0.555556   
2124                   33.0       0.222222       0.222222       0.555556   
2125                   33.0       0.222222       0.222222       0.555556   
2126                   33.0       0.222222       

#### Splitting data into target property and features (input features and output property)

In [6]:
y = data_full['LTC']
Before_Features = data_full.drop(['LTC','Compound'],axis = 1)
print('Shape of dataset before feature selection is: ',Before_Features.shape)
Before_Features

Shape of dataset before feature selection is:  (2128, 272)


mean_EffectiveCoordination  var_EffectiveCoordination  \
0                       5.368858               3.150000e-10   
1                       5.368858               3.150000e-10   
2                       5.368858               3.150000e-10   
3                       5.368858               3.150000e-10   
4                       5.368858               3.150000e-10   
...                          ...                        ...   
2123                   11.956922               8.880000e-16   
2124                   11.956922               8.880000e-16   
2125                   11.956922               8.880000e-16   
2126                   11.956922               8.880000e-16   
2127                   11.956922               8.880000e-16   

      min_EffectiveCoordination  max_EffectiveCoordination  \
0                      5.368858                   5.368858   
1                      5.368858                   5.368858   
2                      5.368858                   5.368858   
3                      5.368858                   5.368858   
4                      5.368858                   5.368858   
...                         ...                        ...   
2123                  11.956922                  11.956922   
2124                  11.956922                  11.956922   
2125                  11.956922                  11.956922   
2126                  11.956922                  11.956922   
2127                  11.956922                  11.956922   

      var_MeanBondLength  min_MeanBondLength  max_MeanBondLength  \
0           1.660000e-11                 1.0                 1.0   
1           1.660000e-11                 1.0                 1.0   
2           1.660000e-11                 1.0                 1.0   
3           1.660000e-11                 1.0                 1.0   
4           1.660000e-11                 1.0                 1.0   
...                  ...                 ...                 ...   
2123        2.220000e-16                 1.0                 1.0   
2124        2.220000e-16                 1.0                 1.0   
2125        2.220000e-16                 1.0                 1.0   
2126        2.220000e-16                 1.0                 1.0   
2127        2.220000e-16                 1.0                 1.0   

      mean_BondLengthVariation  var_BondLengthVariation  \
0                     0.146248             2.090000e-11   
1                     0.146248             2.090000e-11   
2                     0.146248             2.090000e-11   
3                     0.146248             2.090000e-11   
4                     0.146248             2.090000e-11   
...                        ...                      ...   
2123                  0.051982             9.710000e-17   
2124                  0.051982             9.710000e-17   
2125                  0.051982             9.710000e-17   
2126                  0.051982             9.710000e-17   
2127                  0.051982             9.710000e-17   

      min_BondLengthVariation  ...  min_SpaceGroupNumber  \
0                    0.146248  ...                     2   
1                    0.146248  ...                     2   
2                    0.146248  ...                     2   
3                    0.146248  ...                     2   
4                    0.146248  ...                     2   
...                       ...  ...                   ...   
2123                 0.051982  ...                     2   
2124                 0.051982  ...                     2   
2125                 0.051982  ...                     2   
2126                 0.051982  ...                     2   
2127                 0.051982  ...                     2   

      most_SpaceGroupNumber  frac_sValence  frac_pValence  frac_dValence  \
0                     113.5       0.500000       0.500000       0.000000   
1                     113.5       0.500000       0.500000       0.000000   
2                     113.5  

#### Out of 272 features available in dataset, only important features need to be selected.Columns with constant values need to be dropped first. 

In [7]:
## VT Analysis
from sklearn.feature_selection import VarianceThreshold
var_thres = VarianceThreshold(threshold=.8*(1-0.8))
var_thres.fit(Before_Features)
var_thres.get_support()
constant_columns = [column for column in Before_Features.columns if column not in Before_Features.columns[var_thres.get_support()]]
After_Variance = Before_Features.drop(constant_columns,axis=1)
print('Shape data after feature selection is: ',After_Variance.shape)
After_Variance

Shape data after feature selection is:  (2128, 211)


mean_EffectiveCoordination  var_EffectiveCoordination  \
0                       5.368858               3.150000e-10   
1                       5.368858               3.150000e-10   
2                       5.368858               3.150000e-10   
3                       5.368858               3.150000e-10   
4                       5.368858               3.150000e-10   
...                          ...                        ...   
2123                   11.956922               8.880000e-16   
2124                   11.956922               8.880000e-16   
2125                   11.956922               8.880000e-16   
2126                   11.956922               8.880000e-16   
2127                   11.956922               8.880000e-16   

      min_EffectiveCoordination  max_EffectiveCoordination  \
0                      5.368858                   5.368858   
1                      5.368858                   5.368858   
2                      5.368858                   5.368858   
3                      5.368858                   5.368858   
4                      5.368858                   5.368858   
...                         ...                        ...   
2123                  11.956922                  11.956922   
2124                  11.956922                  11.956922   
2125                  11.956922                  11.956922   
2126                  11.956922                  11.956922   
2127                  11.956922                  11.956922   

      mean_NeighDiff_shell1_Number  var_NeighDiff_shell1_Number  \
0                         1.773368                 5.840000e-11   
1                         1.773368                 5.840000e-11   
2                         1.773368                 5.840000e-11   
3                         1.773368                 5.840000e-11   
4                         1.773368                 5.840000e-11   
...                            ...                          ...   
2123                     21.727741                 1.780000e-15   
2124                     21.727741                 1.780000e-15   
2125                     21.727741                 1.780000e-15   
2126                     21.727741                 1.780000e-15   
2127                     21.727741                 1.780000e-15   

      min_NeighDiff_shell1_Number  max_NeighDiff_shell1_Number  \
0                        1.773368                     1.773368   
1                        1.773368                     1.773368   
2                        1.773368                     1.773368   
3                        1.773368                     1.773368   
4                        1.773368                     1.773368   
...                           ...                          ...   
2123                    21.727741                    21.727741   
2124                    21.727741                    21.727741   
2125                    21.727741                    21.727741   
2126                    21.727741                    21.727741   
2127                    21.727741                    21.727741   

      range_NeighDiff_shell1_Number  mean_NeighDiff_shell1_MendeleevNumber  \
0                      1.170000e-10                               8.866839   
1                      1.170000e-10                               8.866839   
2                      1.170000e-10                               8.866839   
3                      1.170000e-10                               8.866839   
4                      1.170000e-10                               8.866839   
...                             ...                                    ...   
2123                   3.550000e-15                              15.519815   
2124                   3.550000e-15                              15.519815   
2125                   3.550000e-15                              15.519815   
2126                   3.550000e-15                              15.519815   
2127                   3.550000e-15             

#### Now, columns with High correlation coefficient need to be dropped. 

In [8]:
## PC Analysis
import matplotlib.pyplot as plt 
import seaborn as sns
def correlation(dataset, threshold):
    col_corr = set()  # Set of all the names of correlated columns
    corr_matrix = dataset.corr()
    for i in range(len(corr_matrix.columns)):
        for j in range(i):
            if abs(corr_matrix.iloc[i, j]) > threshold: # we are interested in absolute coeff value
                colname = corr_matrix.columns[i]  # getting the name of column
                col_corr.add(colname)
    af_corr = dataset.drop(col_corr,axis=1)
    return af_corr

af_both2 = correlation(After_Variance, 0.80)
print('Now,shape of input features is: ',af_both2.shape)
af_both2

Now,shape of input features is:  (2128, 57)


mean_EffectiveCoordination  var_EffectiveCoordination  \
0                       5.368858               3.150000e-10   
1                       5.368858               3.150000e-10   
2                       5.368858               3.150000e-10   
3                       5.368858               3.150000e-10   
4                       5.368858               3.150000e-10   
...                          ...                        ...   
2123                   11.956922               8.880000e-16   
2124                   11.956922               8.880000e-16   
2125                   11.956922               8.880000e-16   
2126                   11.956922               8.880000e-16   
2127                   11.956922               8.880000e-16   

      min_EffectiveCoordination  mean_NeighDiff_shell1_Number  \
0                      5.368858                      1.773368   
1                      5.368858                      1.773368   
2                      5.368858                      1.773368   
3                      5.368858                      1.773368   
4                      5.368858                      1.773368   
...                         ...                           ...   
2123                  11.956922                     21.727741   
2124                  11.956922                     21.727741   
2125                  11.956922                     21.727741   
2126                  11.956922                     21.727741   
2127                  11.956922                     21.727741   

      var_NeighDiff_shell1_Number  mean_NeighDiff_shell1_MendeleevNumber  \
0                    5.840000e-11                               8.866839   
1                    5.840000e-11                               8.866839   
2                    5.840000e-11                               8.866839   
3                    5.840000e-11                               8.866839   
4                    5.840000e-11                               8.866839   
...                           ...                                    ...   
2123                 1.780000e-15                              15.519815   
2124                 1.780000e-15                              15.519815   
2125                 1.780000e-15                              15.519815   
2126                 1.780000e-15                              15.519815   
2127                 1.780000e-15                              15.519815   

      var_NeighDiff_shell1_MendeleevNumber  mean_NeighDiff_shell1_MeltingT  \
0                             2.920000e-10                      546.348032   
1                             2.920000e-10                      546.348032   
2                             2.920000e-10                      546.348032   
3                             2.920000e-10                      546.348032   
4                             2.920000e-10                      546.348032   
...                                    ...                             ...   
2123                          0.000000e+00                      147.554643   
2124                          0.000000e+00                      147.554643   
2125                          0.000000e+00                      147.554643   
2126                          0.000000e+00                      147.554643   
2127                          0.000000e+00                      147.554643   

      var_NeighDiff_shell1_MeltingT  mean_NeighDiff_shell1_CovalentRadius  \
0                      1.800000e-08                             12.413575   
1                      1.800000e-08                             12.413575   
2                      1.800000e-08                             12.413575   
3                      1.800000e-08                             12.413575   
4                      1.800000e-08                             12.413575   
...                             ...                                   ...   
2123                   1.420000e-14                              4.655945   
21

In [40]:
total_data, last_data,total_y,last_y = af_both2[:1900], af_both2[1900:],y[:1900],y[1900:]

#### MinMax Scaling need to be done to speed up convergence of training of Model

In [41]:
##Scaling
from sklearn.preprocessing import MinMaxScaler
scaler = MinMaxScaler()
df_scaled = scaler.fit_transform(total_data)
df_scaled = pd.DataFrame(df_scaled)
df_scaled.columns =af_both2.columns

In [42]:
df_scaled

mean_EffectiveCoordination  var_EffectiveCoordination  \
0                       0.002126               1.339281e-10   
1                       0.002126               1.339281e-10   
2                       0.002126               1.339281e-10   
3                       0.002126               1.339281e-10   
4                       0.002126               1.339281e-10   
...                          ...                        ...   
1895                    0.055344               2.074823e-15   
1896                    0.055344               2.074823e-15   
1897                    0.055344               2.074823e-15   
1898                    0.055344               2.074823e-15   
1899                    0.055344               2.074823e-15   

      min_EffectiveCoordination  mean_NeighDiff_shell1_Number  \
0                      0.011392                      0.004632   
1                      0.011392                      0.004632   
2                      0.011392                      0.004632   
3                      0.011392                      0.004632   
4                      0.011392                      0.004632   
...                         ...                           ...   
1895                   0.064117                      0.440489   
1896                   0.064117                      0.440489   
1897                   0.064117                      0.440489   
1898                   0.064117                      0.440489   
1899                   0.064117                      0.440489   

      var_NeighDiff_shell1_Number  mean_NeighDiff_shell1_MendeleevNumber  \
0                    3.136153e-12                               0.088147   
1                    3.136153e-12                               0.088147   
2                    3.136153e-12                               0.088147   
3                    3.136153e-12                               0.088147   
4                    3.136153e-12                               0.088147   
...                           ...                                    ...   
1895                 6.658956e-16                               0.009109   
1896                 6.658956e-16                               0.009109   
1897                 6.658956e-16                               0.009109   
1898                 6.658956e-16                               0.009109   
1899                 6.658956e-16                               0.009109   

      var_NeighDiff_shell1_MendeleevNumber  mean_NeighDiff_shell1_MeltingT  \
0                             1.286175e-11                        0.248160   
1                             1.286175e-11                        0.248160   
2                             1.286175e-11                        0.248160   
3                             1.286175e-11                        0.248160   
4                             1.286175e-11                        0.248160   
...                                    ...                             ...   
1895                          3.911382e-17                        0.449826   
1896                          3.911382e-17                        0.449826   
1897                          3.911382e-17                        0.449826   
1898                          3.911382e-17                        0.449826   
1899                          3.911382e-17                        0.449826   

      var_NeighDiff_shell1_MeltingT  mean_NeighDiff_shell1_CovalentRadius  \
0                      2.369582e-11                              0.092292   
1                      2.369582e-11                              0.092292   
2                      2.369582e-11                              0.092292   
3                      2.369582e-11                              0.092292   
4                      2.369582e-11                              0.092292   
...                             ...                                   ...   
1895                   6.740144e-16                              0.193119   
18

#### Splitting data into training and testing: 

In [43]:
# converting input data into numpy array format.
input_features = df_scaled.to_numpy()
output_property = total_y.to_numpy()

##Vizualizing the split
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(input_features, output_property, test_size=0.2, random_state=90)
# sns.histplot(data=y,bins=20,kde=True,legend=True)
# plt.show()
# sns.histplot(data=y_train,color = "g",bins=20,kde=True)
# plt.show()
# sns.histplot(data=y_test,color="r",bins=20,kde=True)
# plt.show();

#### Now, Data is ready to model preparation and validation:

In [44]:
##RF Algorithm
from sklearn.ensemble import RandomForestRegressor
rf_reg = RandomForestRegressor()
rf_reg.fit(X_train,y_train)
print('Training Accuracy: {}'.format(rf_reg.score(X_train,y_train)))
print('Test Accuracy: {}'.format(rf_reg.score(X_test,y_test)))

Training Accuracy: 0.9899521784631229
Test Accuracy: 0.9354591834723949


#### Cross-Validation (K-Fold) is performed to check model performance on different random sampling.

In [14]:
##Repeated K-fold

# scores = []
# counter = 0
# from sklearn.model_selection import RepeatedKFold
# fold = RepeatedKFold(n_splits = 10,n_repeats=10, random_state=30)
# for train_index, test_index in fold.split(af_both2):
#     counter = counter + 1
#     X_train, X_test = af_both2.iloc[train_index], af_both2.iloc[test_index]
#     y_train, y_test = y.iloc[train_index], y.iloc[test_index]
#     rf_reg.fit(X_train,y_train)
#     scores.append(rf_reg.score(X_test,y_test))
# print(scores)

In [55]:
rf_reg.predict(total_data[:20].to_numpy())

array([191.3799605, 191.3799605, 191.3799605, 191.3799605, 191.3799605,
       191.3799605, 191.3799605, 191.3799605, 191.3799605, 191.3799605,
       191.3799605, 191.3799605, 191.3799605, 191.3799605, 191.3799605,
       191.3799605, 191.3799605, 191.3799605, 197.5510891, 197.5510891])

In [56]:
total_y[:20]

0     181.81181
1     154.58458
2     127.35736
3     108.29830
4      89.23924
5      75.62563
6      56.56657
7      45.67568
8      40.23023
9      34.78478
10     26.61662
11     23.89389
12     21.17117
13     18.44845
14     15.72573
15     13.00300
16     13.00300
17     10.28028
18    406.56566
19    330.23416
Name: LTC, dtype: float64

In [29]:
# Predicting LTC of material using Features of alloy. 
X = total_data    # Features of Alloy 
print(rf_reg.predict([X,input_features[-1],input_features[-2],input_features[-3]])) # Predicted LTC at some temperature
print(input_features[-2][-1])

[0.0460868 0.0460868 0.0462964 0.0449278]
0.775542732282113


In [18]:
rf_reg.get_metadata_routing()

{'fit': {'sample_weight': None}, 'score': {'sample_weight': None}}

### Prediction Code Analysis: 

In [16]:
import numpy as np
import pandas as pd
import os
import pymatgen.core as pg

In [17]:
##Initializing A,A_1, B and X as elements used in this work. Please refer to Fig. 6a in the paper

A = ["Li","Na","K","Rb","Cs"]
A_1 = ["Be","Mg","Ca","Sr","Ba"]
B = ["Ti","Zr","Hf","V","Nb","Ta","Cr","Mo","W","Mn","Tc","Re","Fe","Os","Ru","Co","Rh","Ir","Ni","Pt",
     "Pd","Cu","Ag","Au","Zn","Cd","Hg","Al","Ga","In","Tl","Si","Ge","Ga","Sn","Pb","As","Sb","Bi","Se","Te"]
X = ["F","Cl","Br","I","O","S","P"]

In [20]:
##Read all 4 prototype 
# structures and replace A,B,X by elements from A,bA_1, B, and X

os.chdir(r"prototype_Poscar")
with open('protoA3B2X9.txt') as f:  
    lines = f.read()

In [21]:
#this is to create the properties file header that will be used in Magpie later

for_text = []
initial = "filename delta_e structure"
for_text.append(initial)

In [29]:
cd ..

/raid/phy_aftab/aftab/piyush/Lattice_RF_ML/am1c17378_si_001/Supplementary files/ML Code


In [49]:
##Make a directory A3B2X9 in prototypes to create poscar files from the elemtens and prototype structure A3B2X9
os.chdir(r"/raid/phy_aftab/aftab/piyush/Lattice_RF_ML/am1c17378_si_001/Supplementary files/ML Code/prototype_Poscar/prototypes/A3B2X9")
counter = 1
for i in A:
    for j in B:
        for k in X:
            
            for_output = lines.replace("A",i)
            for_output_2 = for_output.replace("B",j)
            for_output_3 = for_output_2.replace("X",k)
            for_text.append(f"{counter}-{i}3{j}2{k}9 none {counter}")
            with open(f"{counter}-{i}3{j}2{k}9", "w") as text_file:
                text_file.write(for_output_3)
            counter = counter + 1

In [50]:
##This is for the properties file again
textfile = open("properties.txt", "w")
for element in for_text:
    textfile.write(element + "\n")
textfile.close()